jetty-util/src/main/java/org/eclipse/jetty/util/thread/ReservedThreadExecutor.java

```diff
@@ -266,139 +254,136 @@ public String toString()
                     if (task != null)
                         return task;
 
-                    if (_stack.remove(this))
+                    // Have we idled out?
+                    if (_idleTime > 0)
                     {
-                        if (LOG.isDebugEnabled())
-                            LOG.debug("{} IDLE", this);
-                        _size.decrementAndGet();
+                        if (!_state.compareAndSet(State.POLLING, State.IDLED_OUT))
+                            LOG.warn("Bad idle state: {} in {}", this, ReservedThreadExecutor.this);
+                        else if (LOG.isDebugEnabled())
+                            LOG.debug("Idle {} in {}", this, ReservedThreadExecutor.this);
+                        _count.add(0, -1);
                         return STOP;
                     }
                 }
                 catch (InterruptedException e)
                 {
                     LOG.ignore(e);
-                    // If the wait was interrupted, then STOP if we are not running
-                    if (!isRunning())
-                        return STOP;
                 }
             }
         }
 
         @Override
         public void run()
         {
-            while (isRunning())
+            _thread = Thread.currentThread();
+            try
             {
-                // test and increment size BEFORE decrementing pending,
-                // so that we don't have a race starting new pending.
-                int size = _size.get();
-
-                // Are we stopped?
-                if (size < 0)
-                    return;
-
-                // Are we surplus to capacity?
-                if (size >= _capacity)
-                {
-                    if (LOG.isDebugEnabled())
-                        LOG.debug("{} size {} > capacity", this, size, _capacity);
-                    if (_starting)
-                        _pending.decrementAndGet();
-                    return;
-                }
-
-                // If we cannot update size then recalculate
-                if (!_size.compareAndSet(size, size + 1))
-                    continue;
-
-                if (_starting)
+                while (true)
                 {
+                    boolean exit = false;
+                    long count = _count.get();
+
+                    // check the state
+                    State state = _state.get();
+
+                    // reduce pending if this thread was pending
+                    int pending = AtomicBiInteger.getHi(count) - (state == State.PENDING ? 1 : 0);
+                    int size = AtomicBiInteger.getLo(count);
+
+                    // Exit if excess capacity or it has been too long since we hit zero reserved threads?
+                    if (size >= _capacity ||
+                        size > 0 && _idleTime > 0 && _idleTimeUnit.toNanos(_idleTime) < (_lastZero - System.nanoTime()) && _state.compareAndSet(state, State.IDLED_OUT))
```

Comment:
1. What is that `_lastZero` logic about? On a busy server, I expect the `size` counter to never reach 0 and I wouldn't want to have to re-create the `ReservedThread`s every minute or so, even if that would be fairly cheap, I fail to see the usefulness of such logic.

2. The state change to `IDLED_OUT` here will make the state change to `POLLING` a few lines below always fail because you don't refresh the `state` local variable.

Reply:
1.  If we have not been to 0 reserved threads for a period longer than the idle timeout, then we timeout.
2. Exactly!

Reply:
I'll look to see if this logic can be made clearer...

Reply:
1. Is that useful?
2. `IllegalStateException("Bad State")` is thrown when this happens.

Reply:
1. @lorban The issue I'm trying so solve here is that if we have multiple reserved threads waiting on the SQ, then there is no guarantee of the order in which they wake up.  Let's say wake up is fair and longest waiting wakes up first, then this is bad for idling out reserved threads.   If we have 30 reserved threads, 30s idle timeout and 1 request per second, then each of the 30 reserved threads will receive 1 request every 30 seconds and thus never time out.

2. will fix

Reply:
fixed the ordering and clarified the algorithm a bit.  Also slowed the shrinkage rate